pkg/graveler/graveler.go

```diff
@@ -1606,29 +1629,50 @@ func (g *KVGraveler) List(ctx context.Context, repositoryID RepositoryID, ref Re
 }
 
 func (g *KVGraveler) Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, params CommitParams) (CommitID, error) {
-	// TODO(issue 3569) - kv implementation optimistic lock commit
 	var preRunID string
 	var commit Commit
+	var newCommitID CommitID
 	var storageNamespace StorageNamespace
-	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
-		isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+	if err != nil {
+		return "", err
+	}
+	if isProtected {
+		return "", ErrCommitToProtectedBranch
+	}
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", fmt.Errorf("get repository: %w", err)
+	}
+	storageNamespace = repo.StorageNamespace
+
+	branch, err := g.RefManager.GetBranch(ctx, repositoryID, branchID)
+	if err != nil {
+		return "", fmt.Errorf("get branch: %w", err)
+	}
+
+	if params.SourceMetaRange != nil {
+		empty, err := g.stagingEmpty(ctx, repositoryID, repo, branch)
 		if err != nil {
-			return nil, err
+			return "", fmt.Errorf("checking empty branch: %w", err)
 		}
-		if isProtected {
-			return nil, ErrCommitToProtectedBranch
-		}
-		repo, err := g.GetRepository(ctx, repositoryID)
-		if err != nil {
-			return "", fmt.Errorf("get repository: %w", err)
+		if !empty {
+			return "", ErrCommitMetaRangeDirtyBranch
 		}
-		storageNamespace = repo.StorageNamespace
+	}
 
-		branch, err := g.db.GetBranch(ctx, repositoryID, branchID)
-		if err != nil {
-			return "", fmt.Errorf("get branch: %w", err)
-		}
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		tokens := make([]StagingToken, len(branch.SealedTokens)+1)
+		copy(tokens, branch.SealedTokens)
+		tokens[len(tokens)-1] = branch.StagingToken
+
+		branch.StagingToken = generateStagingToken(repositoryID, branchID)
+		branch.SealedTokens = tokens
```

Comment:
Any reason to clone the data?

```suggestion
		branch.SealedTokens = append(branch.SealedTokens, branch.StagingToken)
		branch.StagingToken = generateStagingToken(repositoryID, branchID)
```

Reply:
done